### ISSUE TITLE
Address spaces and function parameters

### ISSUE THREAD

**Issue by kpet** [KhronosGroup/OpenCL-Docs#644]
The OpenCL C specification currently states (emphasis mine):

>Most of the restrictions from section 5.1.2 and section 5.3 of the Embedded C Specification apply in
OpenCL C, e.g. address spaces can not be used with a return type, **a function parameter**, or a
function type, and multiple address space qualifiers are not allowed. 

Address space qualifiers on function parameters are accepted by compilers and a very necessary feature in the absence of generic address space support. Here's a quick code example:

```
int example(global int* ptr, int val) {
    return *ptr + val;
}
```
Am I misunderstanding what the specification means by "function parameter". Is this a mistake?

**Comment by AnastasiaStulova** [comment]
I think what the spec means is that the address spaces are not accepted as top-level qualifiers in parameters. This is the same as for embedded C. But we should improve the wording to be more specific.

**Comment by AnastasiaStulova** [comment]
I would just like to point out that there is one exception where we accept `__private` address space in those cases because this is what it deduced to according to:

`All function arguments as well as return values are in the private address space.`

**Comment by kpet** [comment]
Talking some more with @AnastasiaStulova (thanks!), what this means in layman terms is that the arguments themselves always are in the private address space but nothing prevents a pointer residing in the private address space to point to another address space as in my example. With that in mind the specification language makes perfect sense. I think a couple of examples of valid and invalid cases might be enough to clarify this for everybody. We may also want to consider using more specific language to define the rules, as suggested by @AnastasiaStulova, as well as explain what the language means to improve readability for general users.

**Comment by alycm** [comment]
I think there is potential for a similar clarification regarding "multiple address space qualifiers are not allowed" as well.  As a single type can have multiple address space qualifiers, just at different levels of pointer indirection.

```c
kernel void foo()
{
    private local int x;  // Illegal
}

kernel void foo(local int *x)
{
    local int * private * xp = &x;  // Legal
}
```

Examples probably are a good way to demonstrate this as we're not even really specifying this any more but inheriting from Embedded C.
